### PR TITLE
gh-104400: pygettext: use an AST parser instead of a tokenizer

### PR DESCRIPTION
**Pull request by tomasr8**
This PR replaces the token-based message extraction with one that uses the AST parser instead.
See the [issue](https://github.com/python/cpython/issues/104400) or the [forum discussion](https://discuss.python.org/t/modernize-and-add-missing-features-to-pygettext/26455) for more info.

This change fixes some issues just by virtue of using AST instead of working directly with tokens:

- docstrings with leading blank lines are extracted correctly
- dosctrings like `"""Hello, {}!""".format('world')` are no longer extracted
- docstrings are cleaned with  `inspect.cleandoc()` via `ast.get_docstring()`
- This is now correctly extracted:
```python
def test(x=_('param')):
    pass
```

~~I added a CLI argument `--charset` (same as in pybabel and `--from-code` in xgettext) to force a file encoding, e.g. `--charset=utf-8` will open the source files with `utf-8` encoding. This is useful because currently we are relying on the system default which is error-prone. For example on Windows, `open()` in my locale defaults to `cp1250` which mangles up `utf-8` files and vice versa (with some `UnicodeDecoreError`s in between).~~
- Let'd do this in a separate PR in order to keep the diff as small as possible here (this is also not needed when running python with `-Xutf8`)

This PR has lots more tests to make sure we don't regress on anything. The tests now compare the script output to a `.po` file rather than just comparing the `msgid`s (basically snapshot tests). This ensures that we also catch issues with formatting, line locations or anything else.

@warsaw if you feel like having a look (or anyone else ;))

* Issue: gh-104400

### CODE REVIEW

File: Tools/i18n/pygettext.py [hunk 1]
```diff
@@ -207,7 +210,7 @@ def make_escapes(pass_nonascii):
     global escapes, escape
     if pass_nonascii:
         # Allow non-ascii characters to pass through so that e.g. 'msgid
-        # "H�he"' would result not result in 'msgid "H\366he"'.  Otherwise we
+        # "Höhe"' would result not result in 'msgid "H\366he"'.  Otherwise we
```
Comment (arhadthedev) [inline Tools/i18n/pygettext.py]: The `would result not result` part is confusing. It should be either `would result` or `would not result` since the PR touches this line anyway.

Comment (arhadthedev) [reply Tools/i18n/pygettext.py]: (I forgot to press the `Submit review` button until I noticed the Pending badge here today)

**Comment by tomasr8** [comment]
@ambv This is what I talked to you about at EuroPython. If you have time I'd be very happy if you could have a look :)

The TL;DR is pygettext has a couple of bugs which stem from it using a tokenizer-based extraction (and overall the code needs modernizing). I fix those bugs in this PR by switching to a parser. Otherwise I try to keep the functionality as close as possible. I also added lots more tests which compare the entire output and not just the messages as it was previously.

There are also lots of features missing in pygettext - handling ngettext, pgettext and others, format flags, etc..
Once this is done, I will submit patches for those missing features as well - I didn't want to put everything in one giant PR as it's already pretty big.

Thank you!

**Comment by AA-Turner** [comment]
@tomasr8 would it be possible to break this PR up into several chunks / stages? You may have more luck with reviewers & progress -- I'm happy to help if wanted.

A

**Comment by tomasr8** [comment]
> @tomasr8 would it be possible to break this PR up into several chunks / stages? You may have more luck with reviewers & progress -- I'm happy to help if wanted.

I can definitely give it a try! I think it'll be difficult to separate the actual change from tokens to AST, since that's kind of an all-or-nothing change but I could start with improving the tests first in a separate PR. That should be an added value regardless of whether the rest gets merged or not. I'll see if I can get a separate PR for the tests in the coming days.

Any help/review is greatly appreciated of course! :)

**Comment by tomasr8** [comment]
@AA-Turner I opened a separate [PR](https://github.com/python/cpython/pull/108173) just adding extra tests, if you wanna have a look ;)

**Comment by Wulian233** [comment]
There was a recent issue in support of f-string https://github.com/python/cpython/issues/113604 , that mentioned this. AST makes this feature easier to add

https://github.com/python/cpython/pull/108173/files already contains the required tests, can you remove them in this PR, so that the diff will be smaller and easier to review

**Comment by tomasr8** [comment]
> There was a recent issue in support of f-string #113604 , that mentioned this. AST makes this feature easier to add
> 
> https://github.com/python/cpython/pull/108173/files already contains the required tests, can you remove them in this PR, so that the diff will be smaller and easier to review

Don't waste your time reviewing this PR just yet, we should get the tests merged before moving on with this :) Actually, it's been a while since I opened the tests PR, it might need updating first..

**Comment by tomasr8** [comment]
Alright, I think this is reviewable again. With https://github.com/python/cpython/pull/126912 merged, it was possible to simplify this PR quite a bit since the data structures are already in place. The main change is really just about replacing the tokenizer with a parser.

Using a parser fixes a lot of issues that would've been hard and/or annoying to fix within the tokenizer (see the first post for a list).

I also added some additional tests for error messages. Feedback welcome!

File: Tools/i18n/pygettext.py [hunk 2]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-            self.__state = self.__ignorenext
+class GettextVisitor(NodeVisitor):
+    def __init__(self, options, filename=None):
+        super().__init__()
+        self.options = options
+        self.filename = filename
+        self.messages = {}
+
+    def visit(self, node):
+        if type(node) in {Module, FunctionDef, AsyncFunctionDef, ClassDef}:
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: Why not use visit_Module etc?

File: Tools/i18n/pygettext.py [hunk 3]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-            elif tstring in ')]}':
-                self.__enclosurecount -= 1
-
-    def __ignorenext(self, ttype, tstring, lineno):
-        self.__state = self.__waiting
-
-    def __addentry(self, msg, lineno=None, *, is_docstring=False):
-        msgid = msg.get('msgid')
-        if msgid in self.__options.toexclude:
+        if (docstring := ast.get_docstring(node, clean=True)) is not None:
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: I do not think that using `:=` makes this code more clear.

**Comment by tomasr8** [comment]
cc @serhiy-storchaka in case you missed this :slightly_smiling_face:

File: Tools/i18n/pygettext.py [hunk 4]
```diff
@@ -155,19 +155,19 @@
...
 import importlib.util
+import os
 import sys
-import glob
 import time
-import getopt
-import ast
-import tokenize
-from collections import defaultdict
+from ast import AsyncFunctionDef, ClassDef, FunctionDef, Module, NodeVisitor
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
```

File: Tools/i18n/pygettext.py [hunk 5]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-                if ttype in (tokenize.COMMENT, tokenize.NL, tokenize.ENCODING):
-                    return
-                self.__freshmodule = 0
-            # class or func/method docstring?
-            if ttype == tokenize.NAME and tstring in ('class', 'def'):
-                self.__state = self.__suiteseen
-                return
-        if ttype == tokenize.NAME and tstring in ('class', 'def'):
-            self.__state = self.__ignorenext
+class GettextVisitor(NodeVisitor):
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
class GettextVisitor(ast.NodeVisitor):
```

File: Tools/i18n/pygettext.py [hunk 6]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+    def __init__(self, options, filename=None):
+        super().__init__()
+        self.options = options
+        self.filename = filename
+        self.messages = {}
+
+    def visit(self, node):
+        if type(node) in {Module, FunctionDef, AsyncFunctionDef, ClassDef}:
+            self._extract_docstring(node)
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
        if isinstance(
            node,
            (ast.Module, ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)
        ):
            self._extract_docstring(node)
```

However, you could also use try/except, as `ast.get_docstring()` has an internal check identical to this:

```suggestion
        try:
            self._extract_docstring(node)
        except TypeError:
            pass
```

File: Tools/i18n/pygettext.py [hunk 7]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-    def __addentry(self, msg, lineno=None, *, is_docstring=False):
-        msgid = msg.get('msgid')
-        if msgid in self.__options.toexclude:
+        if (docstring := ast.get_docstring(node, clean=True)) is not None:
+            lineno = node.body[0].lineno
+            self._add_message(lineno, docstring, is_docstring=True)
+
+    def _extract_message(self, node):
+        funcname = self._get_funcname(node)
+        if (spec := self.options.keywords.get(funcname)) is None:
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]: I generally only use assignment expressions when the variable is only used within the relevant block.
```suggestion
        func_name = self._get_func_name(node)
        spec = self.options.keywords.get(func_name)
        if spec is None:
```

File: Tools/i18n/pygettext.py [hunk 8]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-                        else:
-                            print(locline, file=fp)
-                            locline = f'#:{s}'
-                    if len(locline) > 2:
+    def _key_for(self, msgid, msgctxt=None):
+        if msgctxt is not None:
+            return (msgctxt, msgid)
+        return msgid
+
+    def _get_funcname(self, node):
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
    def _get_func_name(self, node):
```

File: Tools/i18n/pygettext.py [hunk 9]
```diff
@@ -347,273 +339,154 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-                    locline = '#:'
-                    for location in locations:
-                        s = f' {location.filename}:{location.lineno}'
-                        if len(locline) + len(s) <= options.width:
-                            locline = locline + s
-                        else:
-                            print(locline, file=fp)
-                            locline = f'#:{s}'
-                    if len(locline) > 2:
+    def _key_for(self, msgid, msgctxt=None):
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
    @staticmethod
    def _key_for(msgid, msgctxt=None):
```

File: Tools/i18n/pygettext.py [hunk 10]
```diff
@@ -748,15 +621,8 @@ class Options:
...
-            try:
-                tokens = tokenize.tokenize(fp.readline)
-                for _token in tokens:
-                    eater(*_token)
-            except tokenize.TokenError as e:
-                print('%s: %s, line %d, column %d' % (
-                    e.args[0], filename, e.args[1][0], e.args[1][1]),
-                    file=sys.stderr)
+            visitor.filename = filename
+            visitor.visit(ast.parse(fp.read()))
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
            module_tree = ast.parse(fp.read())
            visitor.filename = filename
            visitor.visit(module_tree)
```

**Review by AA-Turner — CHANGES_REQUESTED** [review]
General points: 

* Please adhere to PEP 8 (several lines here exceed 79 characters).
* There are a few changes that increase the diff and could be extracted to a preparatory PR -- extracting `write_pot_file()` to a module-level function, moving `key_for` to the class, sorting imports, adding extra whitespace.
* We could consider defining the public interface of `pygettext` to just the command-line interface; otherwise, this PR does affect non-underscored names

A

**Comment by tomasr8** [comment]
Thanks a lot for the review!

> There are a few changes that increase the diff and could be extracted to a preparatory PR -- extracting write_pot_file() to a module-level function, moving key_for to the class, sorting imports, adding extra whitespace.

Good point! I'll do that :)

> We could consider defining the public interface of pygettext to just the command-line interface

Yes I think that makes sense, it is a CLI tool after all (eventually some of the functionality such as writing POT files could be made available but I think that is a long way off). How do we go about doing that? A note in the docs?

**Comment by AA-Turner** [comment]
> How do we go about doing that? A note in the docs?

I don't think `pygettext` is documented anywhere beyond a passing reference in the `gettext` module (it is in `Tools/`, after all). A note in the docstring would probably suffice, or a comment at the top of the file?

A

**Comment by tomasr8** [comment]
Yeah, it doesn't have a dedicated docs page. The module dosctring seems like the best choice

File: Tools/i18n/pygettext.py
Comment (m-aciek) [inline Tools/i18n/pygettext.py:1]: Introduction in docstring with a statement about GNU Gettext is no longer up-to-date, as far as I'm concerned. I believe nowadays it extracts from Python code ~just fine. Would you be willing to update it at the occasion?

```
There's one problem though; xgettext is the program that scans source code
looking for message strings, but it groks only C (or C++). Python
introduces a few wrinkles, such as dual quoting characters, triple quoted
strings, and raw strings. xgettext understands none of this.
```

Comment (tomasr8) [reply Tools/i18n/pygettext.py:1]: Yeah, you're right, xgettext supports Python these days. I will update the docstring :+1:

**Comment by tomasr8** [comment]
Thank you once again for the review! I think this can be reviewed again :) I believe I addressed all your points. I also fixed the lines to not exceed 79 characters, however there are more long lines in parts of the script which this PR does not touch. Do you want me to fix them as well or would you prefer to keep the diff smaller?

I have made the requested changes; please review again :slightly_smiling_face:

File: Lib/test/test_tools/i18n_data/messages.py [hunk 11]
```diff
@@ -32,23 +33,22 @@
 _(None)
 _(1)
 _(False)
-_(("invalid"))
```
Comment (AA-Turner) [inline Lib/test/test_tools/i18n_data/messages.py:35]: Perhaps this was intended to be a one-tuple? The following elements are a single-element list and set.

```suggestion
_(("invalid",))
```

File: Tools/i18n/pygettext.py [hunk 12]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-                    return
-                self.__freshmodule = 0
-            # class or func/method docstring?
-            if ttype == tokenize.NAME and tstring in ('class', 'def'):
-                self.__state = self.__suiteseen
-                return
-        if ttype == tokenize.NAME and tstring in ('class', 'def'):
-            self.__state = self.__ignorenext
+class GettextVisitor(ast.NodeVisitor):
+    def __init__(self, options, filename=None):
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]:
```suggestion
    def __init__(self, options):
```

File: Tools/i18n/pygettext.py [hunk 13]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+    def _extract_docstring(self, node):
+        if (not self.options.docstrings or
+            self.options.nodocstrings.get(self.filename)):
             return
-        if ttype == tokenize.NAME and tstring in opts.keywords:
-            self.__state = self.__keywordseen
-            self.__curr_keyword = tstring
+
+        if (docstring := ast.get_docstring(node)) is not None:
+            lineno = node.body[0].lineno
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]: Could you add a comment to explain which node the line number comes from?

File: Tools/i18n/pygettext.py [hunk 14]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-            if ttype == tokenize.NAME and tstring in ('class', 'def'):
-                self.__state = self.__suiteseen
-                return
-        if ttype == tokenize.NAME and tstring in ('class', 'def'):
-            self.__state = self.__ignorenext
+class GettextVisitor(ast.NodeVisitor):
+    def __init__(self, options, filename=None):
+        super().__init__()
+        self.options = options
+        self.filename = filename
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py]: The filename argument is never used
```suggestion
        self.filename: str = ''
```

File: Tools/i18n/pygettext.py [hunk 15]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+        self.messages = {}
```
Comment (AA-Turner) [inline Tools/i18n/pygettext.py:317]: Add a type annotation?

**Review by AA-Turner — COMMENTED** [review]
I'm not an amazing fan of reusing the same visitor for every file. There's not massive performance benefits, and it means we have tricks with resetting the filename, etc.

However, if you'd prefer to keep the current design, could we add a `walkabout` or `initiate_visit` or etc method that takes the node tree and filename, to conceptually keep the two together during tree traversal?

File: Tools/i18n/pygettext.py [hunk 16]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+
+    def _extract_docstring(self, node):
+        if (not self.options.docstrings or
+            self.options.nodocstrings.get(self.filename)):
             return
-        if ttype == tokenize.NAME and tstring in opts.keywords:
-            self.__state = self.__keywordseen
-            self.__curr_keyword = tstring
+
+        if (docstring := ast.get_docstring(node)) is not None:
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: Please do not use `:=` without need. It does not improve readability.

File: Tools/i18n/pygettext.py [hunk 17]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+class GettextVisitor(ast.NodeVisitor):
+    def __init__(self, options, filename=None):
+        super().__init__()
+        self.options = options
+        self.filename = filename
+        self.messages = {}
+
+    def visit(self, node):
+        try:
+            self._extract_docstring(node)
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: Should not it only be called in `visit_Module`?

File: Tools/i18n/pygettext.py [hunk 18]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+                      f'constant for argument {position + 1}, '
+                      f'got {ast.unparse(arg)}', file=sys.stderr)
                 return
-        if lineno is None:
-            lineno = self.__lineno
-        msgctxt = msg.get('msgctxt')
-        msgid_plural = msg.get('msgid_plural')
+            msg_data[arg_type] = arg.value
+
+        if not matches_spec(msg_data, spec):
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: This is no longer needed, because we iterated all `spec` items above.

File: Tools/i18n/pygettext.py [hunk 19]
```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
-            elif tstring in ')]}':
-                self.__enclosurecount -= 1
-
-    def __ignorenext(self, ttype, tstring, lineno):
-        self.__state = self.__waiting
-
-    def __addentry(self, msg, lineno=None, *, is_docstring=False):
-        msgid = msg.get('msgid')
-        if msgid in self.__options.toexclude:
+        if max(spec) >= len(node.args):
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: Note that this (and the code in the loop below) does not work well for `dgettext(*a, "msg")`. We cannot statically determine the number of arguments before "msg". We should check that there are no var-positional arguments before `max(spec)`.

File: Tools/i18n/pygettext.py [hunk 20]
```diff
@@ -730,15 +595,9 @@ class Options:
...
-            eater.set_filename(filename)
-            try:
-                tokens = tokenize.tokenize(fp.readline)
-                for _token in tokens:
-                    eater(*_token)
-            except tokenize.TokenError as e:
-                print('%s: %s, line %d, column %d' % (
-                    e.args[0], filename, e.args[1][0], e.args[1][1]),
-                    file=sys.stderr)
+            module_tree = ast.parse(fp.read())
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: Since we need all file content at once instead of reading it line by line, we can simplify reading. Read it by
```
source = sys.stdin.buffer.read()
```
and
```
with open(filename, 'rb') as fp:
    source = fp.read()
```
and get rid of `closep` and `try`/`finally`.

File: Tools/i18n/pygettext.py (hunk 15 above)
Comment (serhiy-storchaka) [reply Tools/i18n/pygettext.py:317]: Since type annotations were not used in this file, this is not needed. It will introduce a lot of unrelated changes.

File: Lib/test/test_tools/i18n_data/messages.py (hunk 11 above)
Comment (serhiy-storchaka) [reply Lib/test/test_tools/i18n_data/messages.py:35]: No, it tested a literal string nested in parentheses. Now this is supported.

**Review by serhiy-storchaka — COMMENTED** [review]
Great! It is now much simpler. And it works more correctly!

There are still some issues with var-positional arguments. Since we cannot determine the argument position in such case, it should be rejected.

File: Tools/i18n/pygettext.py (hunk 17 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: Adam pointed out that `ast.get_docstring` internally checks the type of the node being passed, so this is safe to call on an arbitrary node

Comment (serhiy-storchaka) [reply Tools/i18n/pygettext.py]: But this is ineffective.

Comment (tomasr8) [reply Tools/i18n/pygettext.py]: ok, I added a visit method for each node that can contain a docstring :)

File: Tools/i18n/pygettext.py (hunk 18 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: Removed!

File: Tools/i18n/pygettext.py (hunk 19 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: I added an explicit check for var-positional arguments

File: Tools/i18n/pygettext.py (hunk 20 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: Done :)

**Comment by tomasr8** [comment]
> I'm not an amazing fan of reusing the same visitor for every file. There's not massive performance benefits, and it means we have tricks with resetting the filename, etc.

I didn't really do this for performance but so that I don't need to pass the `messages` dictionary. If I create a new visitor for each file, I still need to pass the messages to it. I thought that passing the filename was simpler, so I did it this way, but if you prefer it the other way, just tell me and I'll change it :slightly_smiling_face:  For now, I added `GettextVisitor.visit_file(module_tree, filename)` which sets the filename before extracting the messages.

File: Tools/i18n/pygettext.py [hunk 21]
```diff
@@ -325,203 +309,100 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+
+    def visit_file(self, node, filename):
+        self.filename = filename
+        self.visit(node)
+
+    def visit_Module(self, node):
+        self._extract_docstring(node)
+        self.generic_visit(node)
+
+    def visit_ClassDef(self, node):
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: You can also use aliases if you prefer.
```py
visit_ClassDef = ... = visit_Module
```

File: Tools/i18n/pygettext.py [hunk 22]
```diff
@@ -325,203 +309,100 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
...
+            self, lineno, msgid, msgid_plural=None, msgctxt=None, *,
+            is_docstring=False):
+        if msgid in self.options.toexclude:
+            return
+
         key = self._key_for(msgid, msgctxt)
-        if key in self.__messages:
-            self.__messages[key].add_location(
-                self.__curfile,
+        if message := self.messages.get(key):
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: `:=` is not needed here too.

**Comment by tomasr8** [comment]
(The Windows failure is some unrelated build failure)

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by tomasr8** [comment]
Woohoo! 🎉🎉 it took almost two years but we got there :) Thanks @serhiy-storchaka and @AA-Turner for your patience and help!
